// @mantine/spotlight/cjs/index.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var spotlight_store = require('./spotlight.store.js');
var isActionsGroup = require('./is-actions-group.js');
var Spotlight = require('./Spotlight.js');
var SpotlightAction = require('./SpotlightAction.js');
var SpotlightActionsGroup = require('./SpotlightActionsGroup.js');
var SpotlightActionsList = require('./SpotlightActionsList.js');
var SpotlightEmpty = require('./SpotlightEmpty.js');
var SpotlightFooter = require('./SpotlightFooter.js');
var SpotlightSearch = require('./SpotlightSearch.js');



exports.closeSpotlight = spotlight_store.closeSpotlight;
exports.createSpotlight = spotlight_store.createSpotlight;
exports.createSpotlightStore = spotlight_store.createSpotlightStore;
exports.openSpotlight = spotlight_store.openSpotlight;
exports.spotlight = spotlight_store.spotlight;
exports.toggleSpotlight = spotlight_store.toggleSpotlight;
exports.useSpotlight = spotlight_store.useSpotlight;
exports.isActionsGroup = isActionsGroup.isActionsGroup;
exports.Spotlight = Spotlight.Spotlight;
exports.SpotlightAction = SpotlightAction.SpotlightAction;
exports.SpotlightActionsGroup = SpotlightActionsGroup.SpotlightActionsGroup;
exports.SpotlightActionsList = SpotlightActionsList.SpotlightActionsList;
exports.SpotlightEmpty = SpotlightEmpty.SpotlightEmpty;
exports.SpotlightFooter = SpotlightFooter.SpotlightFooter;
exports.SpotlightSearch = SpotlightSearch.SpotlightSearch;
