// @mantine/spotlight/cjs/SpotlightRoot.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var core = require('@mantine/core');
var hooks = require('@mantine/hooks');
var Spotlight_context = require('./Spotlight.context.js');
var spotlight_store = require('./spotlight.store.js');
var getHotkeys = require('./get-hotkeys.js');
var Spotlight_module = require('./Spotlight.module.css.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e['default'] : e; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const defaultProps = {
  size: 600,
  yOffset: 80,
  zIndex: core.getDefaultZIndex("max"),
  overlayProps: { backgroundOpacity: 0.35, blur: 7 },
  transitionProps: { duration: 200, transition: "pop" },
  store: spotlight_store.spotlightStore,
  clearQueryOnClose: true,
  closeOnActionTrigger: true,
  shortcut: "mod + K",
  maxHeight: 400,
  scrollable: false
};
const SpotlightRoot = core.factory((_props, ref) => {
  const props = core.useProps("SpotlightRoot", defaultProps, _props);
  const _a = props, {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    store,
    children,
    query,
    onQueryChange,
    transitionProps,
    clearQueryOnClose,
    shortcut,
    tagsToIgnore,
    triggerOnContentEditable,
    disabled,
    onSpotlightOpen,
    onSpotlightClose,
    forceOpened,
    closeOnActionTrigger,
    maxHeight,
    scrollable
  } = _a, others = __objRest(_a, [
    "classNames",
    "className",
    "style",
    "styles",
    "unstyled",
    "vars",
    "store",
    "children",
    "query",
    "onQueryChange",
    "transitionProps",
    "clearQueryOnClose",
    "shortcut",
    "tagsToIgnore",
    "triggerOnContentEditable",
    "disabled",
    "onSpotlightOpen",
    "onSpotlightClose",
    "forceOpened",
    "closeOnActionTrigger",
    "maxHeight",
    "scrollable"
  ]);
  const theme = core.useMantineTheme();
  const { opened, query: storeQuery } = spotlight_store.useSpotlight(store);
  const _query = query || storeQuery;
  const setQuery = (q) => {
    onQueryChange == null ? void 0 : onQueryChange(q);
    spotlight_store.spotlightActions.setQuery(q, store);
  };
  const getStyles = core.useStyles({
    name: "Spotlight",
    classes: Spotlight_module['default'],
    props,
    className,
    style,
    classNames,
    styles,
    unstyled
  });
  hooks.useHotkeys(getHotkeys.getHotkeys(shortcut, store), tagsToIgnore, triggerOnContentEditable);
  hooks.useDidUpdate(() => {
    opened ? onSpotlightOpen == null ? void 0 : onSpotlightOpen() : onSpotlightClose == null ? void 0 : onSpotlightClose();
  }, [opened]);
  if (disabled) {
    return null;
  }
  return /* @__PURE__ */ React__default.createElement(
    Spotlight_context.SpotlightProvider,
    {
      value: {
        getStyles,
        query: _query,
        setQuery,
        store,
        closeOnActionTrigger
      }
    },
    /* @__PURE__ */ React__default.createElement(
      core.Modal,
      __spreadProps(__spreadValues({
        ref
      }, others), {
        withCloseButton: false,
        opened: opened || !!forceOpened,
        padding: 0,
        onClose: () => spotlight_store.spotlightActions.close(store),
        className,
        style,
        classNames: core.resolveClassNames({
          theme,
          classNames: [Spotlight_module['default'], classNames],
          props,
          stylesCtx: void 0
        }),
        styles: core.resolveStyles({ theme, styles, props, stylesCtx: void 0 }),
        transitionProps: __spreadProps(__spreadValues({}, transitionProps), {
          onExited: () => {
            var _a2;
            clearQueryOnClose && setQuery("");
            spotlight_store.spotlightActions.clearSpotlightState({ clearQuery: clearQueryOnClose }, store);
            (_a2 = transitionProps == null ? void 0 : transitionProps.onExited) == null ? void 0 : _a2.call(transitionProps);
          }
        }),
        __vars: { "--spotlight-max-height": scrollable ? core.rem(maxHeight) : void 0 },
        __staticSelector: "Spotlight",
        "data-scrollable": scrollable || void 0
      }),
      children
    )
  );
});
SpotlightRoot.classes = Spotlight_module['default'];
SpotlightRoot.displayName = "@mantine/spotlight/SpotlightRoot";

exports.SpotlightRoot = SpotlightRoot;
