// @mantine/spotlight/cjs/SpotlightAction.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var core = require('@mantine/core');
var Spotlight_module = require('./Spotlight.module.css.js');
var Spotlight_context = require('./Spotlight.context.js');
var spotlight_store = require('./spotlight.store.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e['default'] : e; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const defaultProps = {
  dimmedSections: true,
  highlightQuery: false
};
const SpotlightAction = core.factory((_props, ref) => {
  const props = core.useProps("SpotlightAction", defaultProps, _props);
  const _a = props, {
    className,
    style,
    classNames,
    styles,
    id,
    description,
    label,
    leftSection,
    rightSection,
    children,
    dimmedSections,
    highlightQuery,
    highlightColor,
    closeSpotlightOnTrigger,
    onClick,
    onMouseDown,
    keywords,
    vars
  } = _a, others = __objRest(_a, [
    "className",
    "style",
    "classNames",
    "styles",
    "id",
    "description",
    "label",
    "leftSection",
    "rightSection",
    "children",
    "dimmedSections",
    "highlightQuery",
    "highlightColor",
    "closeSpotlightOnTrigger",
    "onClick",
    "onMouseDown",
    "keywords",
    "vars"
  ]);
  const ctx = Spotlight_context.useSpotlightContext();
  const stylesApi = { classNames, styles };
  const labelNode = highlightQuery && typeof label === "string" ? /* @__PURE__ */ React__default.createElement(
    core.Highlight,
    __spreadValues({
      component: "span",
      highlight: ctx.query,
      color: highlightColor
    }, ctx.getStyles("actionLabel", stylesApi)),
    label
  ) : /* @__PURE__ */ React__default.createElement("span", __spreadValues({}, ctx.getStyles("actionLabel", stylesApi)), label);
  return /* @__PURE__ */ React__default.createElement(
    core.UnstyledButton,
    __spreadProps(__spreadValues(__spreadValues({
      ref,
      "data-action": true
    }, ctx.getStyles("action", __spreadValues({ className, style }, stylesApi))), others), {
      onMouseDown: (event) => {
        event.preventDefault();
        onMouseDown == null ? void 0 : onMouseDown(event);
      },
      onClick: (event) => {
        onClick == null ? void 0 : onClick(event);
        if (closeSpotlightOnTrigger != null ? closeSpotlightOnTrigger : ctx.closeOnActionTrigger) {
          spotlight_store.spotlightActions.close(ctx.store);
        }
      },
      tabIndex: -1
    }),
    children || /* @__PURE__ */ React__default.createElement(React__default.Fragment, null, leftSection && /* @__PURE__ */ React__default.createElement(
      core.Box,
      __spreadValues({
        component: "span",
        mod: { position: "left", dimmed: dimmedSections }
      }, ctx.getStyles("actionSection", stylesApi)),
      leftSection
    ), /* @__PURE__ */ React__default.createElement("span", __spreadValues({}, ctx.getStyles("actionBody", stylesApi)), labelNode, /* @__PURE__ */ React__default.createElement("span", __spreadValues({}, ctx.getStyles("actionDescription", stylesApi)), description)), rightSection && /* @__PURE__ */ React__default.createElement(
      core.Box,
      __spreadValues({
        component: "span",
        mod: { position: "right", dimmed: dimmedSections }
      }, ctx.getStyles("actionSection", stylesApi)),
      rightSection
    ))
  );
});
SpotlightAction.classes = Spotlight_module['default'];
SpotlightAction.displayName = "@mantine/spotlight/SpotlightAction";

exports.SpotlightAction = SpotlightAction;
