// @mantine/spotlight/esm/SpotlightSearch.js
import React from 'react';
import { factory, useProps, Input } from '@mantine/core';
import classes from './Spotlight.module.css.js';
import { useSpotlightContext } from './Spotlight.context.js';
import { spotlightActions } from './spotlight.store.js';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const defaultProps = {
  size: "lg"
};
const SpotlightSearch = factory((props, ref) => {
  const _a = useProps(
    "SpotlightSearch",
    defaultProps,
    props
  ), { classNames, styles, onKeyDown, onChange, vars, value } = _a, others = __objRest(_a, ["classNames", "styles", "onKeyDown", "onChange", "vars", "value"]);
  const ctx = useSpotlightContext();
  const inputStyles = ctx.getStyles("search");
  const handleKeyDown = (event) => {
    onKeyDown == null ? void 0 : onKeyDown(event);
    if (event.nativeEvent.code === "ArrowDown") {
      event.preventDefault();
      spotlightActions.selectNextAction(ctx.store);
    }
    if (event.nativeEvent.code === "ArrowUp") {
      event.preventDefault();
      spotlightActions.selectPreviousAction(ctx.store);
    }
    if (event.nativeEvent.code === "Enter") {
      event.preventDefault();
      spotlightActions.triggerSelectedAction(ctx.store);
    }
  };
  return /* @__PURE__ */ React.createElement(
    Input,
    __spreadProps(__spreadValues({
      ref,
      classNames: [{ input: inputStyles.className }, classNames],
      styles: [{ input: inputStyles.style }, styles]
    }, others), {
      value: value != null ? value : ctx.query,
      onChange: (event) => {
        ctx.setQuery(event.currentTarget.value);
        onChange == null ? void 0 : onChange(event);
      },
      onKeyDown: handleKeyDown
    })
  );
});
SpotlightSearch.classes = classes;
SpotlightSearch.displayName = "@mantine/spotlight/SpotlightSearch";

export { SpotlightSearch };
